Content.Server/GameObjects/Components/Construction/ConstructionComponent.cs

```diff
@@ -1,49 +1,547 @@
+                        {
+                            if(edge.Steps.Count > 1)
+                                Edge = edge;
+                            return true;
+                        }
+                        break;
+
+                    case NestedConstructionGraphStep nestedStep:
+                        throw new IndexOutOfRangeException($"Nested construction step not supported as the first step in an edge! Graph: {GraphPrototype.ID} Node: {Node.Name} Edge: {edge.Target}");
+                }
+            }
+
+            return false;
+        }
+
+        private async Task<bool> HandleStep(InteractUsingEventArgs eventArgs, ConstructionGraphEdge? edge = null, ConstructionGraphStep? step = null, bool nested = false)
+        {
+            edge ??= Edge;
+            step ??= edge?.Steps[EdgeStep];
+
+            if (edge == null || step == null)
+                return false;
+
+            foreach (var condition in edge.Conditions)
+            {
+                if (!await condition.Condition(Owner)) return false;
+            }
+
+            var handled = false;
+
+            var doAfterSystem = EntitySystem.Get<DoAfterSystem>();
+
+            var doAfterArgs = new DoAfterEventArgs(eventArgs.User, step.DoAfter, default, eventArgs.Target)
+            {
+                BreakOnDamage = false,
+                BreakOnStun = true,
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                NeedHand = true,
+            };
+
+            var audioSystem = EntitySystem.Get<AudioSystem>();
+
+            switch (step)
+            {
+                case ToolConstructionGraphStep toolStep:
+                    // Gotta take welder fuel into consideration.
+                    if (toolStep.Tool == ToolQuality.Welding)
+                    {
+                        if (eventArgs.Using.TryGetComponent(out WelderComponent? welder) &&
+                            await welder.UseTool(eventArgs.User, Owner, step.DoAfter, toolStep.Tool, toolStep.Fuel))
+                        {
+                            handled = true;
+                        }
+                        break;
+                    }
+
+                    if (eventArgs.Using.TryGetComponent(out ToolComponent? tool) &&
+                        await tool.UseTool(eventArgs.User, Owner, step.DoAfter, toolStep.Tool))
+                    {
+                        handled = true;
+                    }
+                    break;
+
+                // To prevent too much code duplication.
+                case EntityInsertConstructionGraphStep insertStep:
+                    var valid = false;
+                    var entityUsing = eventArgs.Using;
+
+                    switch (insertStep)
+                    {
+                        case PrototypeConstructionGraphStep prototypeStep:
+                            if (eventArgs.Using.Prototype?.ID == prototypeStep.Prototype
```

Comment:
Maybe switch to EntityValid for these ones?

Reply:
done